server/src/main/java/org/elasticsearch/indices/recovery/MultiFileWriter.java

```diff
@@ -179,17 +180,22 @@ public void renameAllTempFiles() throws IOException {
         store.renameTempFilesSafe(tempFileNames);
     }
 
-    static final class FileChunk {
+    private static final class FileChunk implements Releasable {
```

Comment:
I think we would leak these if a recovery fails: `MultiFileWriter#closeInternal` simply drops its `FileChunkWriter` objects on the floor.

Reply:
Right thanks for catching this! I pushed 19517e4b2b4 which I think should be fine since we never get into a state where we don't have a writing thread running but pending chunks in the queue.

Reply:
I don't follow the reasoning here. A recovery can fail for reasons external to `writeFile()`, e.g. the target node leaves the cluster and the shard is reassigned elsewhere.

Reply:
The reasoning is this:

We increment the ref count for the bytes when we create the `FileChunk` which we then pass directly to `writeChunk` No matter what happens externally one of two things can happen as a result in `writeChunk`. Either we put the chunk in the queue to drain or we write it out right away.
If we write it out right away then that will just fail and the reference will be decremented. If we put it in the queue, then whatever thread that is currently doing the writing will fail during its current write and as a result drain the queue and release all the chunks that way.
=> since we always have a thread doing work in `org.elasticsearch.indices.recovery.MultiFileWriter.FileChunkWriter#writeChunk` we can be sure to never fail to release a queued chunk if we drain on exceptions.

Reply:
We discussed this in another channel: the missing piece was that we sometimes receive these chunks out-of-order and may never fill in a gap if the chunk gets lost and then the recovery fails.

Reply:
Alright I pushed https://github.com/elastic/elasticsearch/pull/65921/commits/da56e8355ece729bd40c8828da8722b8f93409c9 here now, should be good for another review :) Thanks!